File: java/core/src/main/java/eu/mais_h/sync/DeserializedDifference.java
package eu.mais_h.sync;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import eu.mais_h.sync.serialize.Deserializer;

class DeserializedDifference<T> implements Difference<T> {

  private final Set<T> added;
  private final Set<T> removed;

  DeserializedDifference(Difference<byte[]> serialized, Deserializer<T> deserializer) {
    added = deserialize(serialized.added(), deserializer);
    removed = deserialize(serialized.removed(), deserializer);
  }

  @Override
  public Set<T> added() {
    return added;
  }

  @Override
  public Set<T> removed() {
    return removed;
  }

  private Set<T> deserialize(Set<byte[]> serialized, Deserializer<T> deserializer) {
    Set<T> deserialized = new HashSet<T>(serialized.size());
    for (byte[] content : serialized) {
      deserialized.add(deserializer.deserialize(content));
    }
    return Collections.unmodifiableSet(deserialized);
  }
}

File: ios/core/core/DeserializedDifference.m
//
//  DeserializedDifference.m
//  core
//
//  Created by Corinne Krych on 12/16/13.
//  Copyright (c) 2013 3musket33rs. All rights reserved.
//

#import "DeserializedDifference.h"
#import "Serialize.h"

@implementation DeserializedDifference
@synthesize added = _added;
@synthesize removed = _removed;

-(id)initWithSerializedDifference:(id<Difference>)serializedDifference deserializer:(id<Deserializer>)deserializer {
    if(self = [super init]) {
        _added = [self deserialize:serializedDifference.added with:deserializer];
        _removed = [self deserialize:serializedDifference.removed with:deserializer];
    }
    return self;
}


//=========================================================================================
// Private Methods
//=========================================================================================
-(NSSet*) deserialize:(NSSet*)serializedSet with:(id<Deserializer>)deserializer {
    NSMutableSet* buildingSet = [NSMutableSet set];
    for(NSData* content in serializedSet) {
        [buildingSet addObject:[deserializer deserialize:content]];
    }
    return [buildingSet copy];
}
@end
